src/lib/auth/authconf.rb

```diff
@@ -0,0 +1,1150 @@
+            else
+                Pam.Remove('mkhomedir')
+            end
+            fix_pam
+            # Start/stop daemons
+            if @autofs_enabled || @sssd_nss.include?('automount') || @ldap_nss.include?('automount')
+                service_enable_start('autofs')
+            else
+                service_disable_stop('autofs')
+            end
+            if @nscd_enabled
+                service_enable_start('nscd')
+            else
+                service_disable_stop('nscd')
+            end
+        end
+        
+        # Create a temporary file holding smb.conf for the specified AD domain. Return absolute path to the temporary file.
+        def ad_create_tmp_smb_conf(ad_domain_name, workgroup_name)
+            out = Tempfile.new("tempfile")
+            out.write("
+[global]
+    security = ads
+    realm = #{ad_domain_name}
+    workgroup = #{workgroup_name}
+    log file = /var/log/samba/%m.log
+    kerberos method = secrets and keytab
+    client signing = yes
+    client use spnego = yes
+")
+            out.close
+            return out
+        end
+
+        # Call package installer to install Samba if it has not yet been installed.
+        def ad_install_samba
+            if Package.Installed('samba') || Package.DoInstall(['samba'])
+                return true
+            end
+            Report.Error(_('Failed to install Samba, required by Active Directory operations.'))
+            return false
+        end
+
+        # Return workgroup name of the given AD domain, return empty string if there is an error.
+        def ad_get_workgroup_name(ad_host_or_domain)
+            if !ad_install_samba
+                return ''
+            end
+            out, status = Open3.capture2("net ads lookup -S #{ad_host_or_domain}")
+            if status.exitstatus != 0
+                return ''
+            end
+            # Look for pre-win2k domain (i.e. workgroup) among the key-value output
+            workgroup_name = ''
+            out.split("\n").each { |line|
+                fields = line.split(':')
+                if fields[0] && fields[0].strip.downcase == 'pre-win2k domain'
+                    workgroup_name = fields[1].strip
+                end
+            }
+            return workgroup_name
+        end
+
+        # Return a tuple of two booleans: AD has this computer entry or not, Kerberos has the appropriate keytab or not.
+        def ad_get_membership_status(ad_domain_name)
+            if !ad_install_samba
+                return [false, false]
+            end
+            smb_conf = ad_create_tmp_smb_conf(ad_domain_name, ad_get_workgroup_name(ad_domain_name))
+            _, status = Open3.capture2("net -s #{smb_conf.path} ads testjoin")
+
+            ad_has_computer = status.exitstatus == 0
+            klist, status = Open3.capture2("klist -k")
+            kerberos_has_key = klist.split("\n").any?{ |line| /#{Socket.gethostname}.*#{ad_domain_name.downcase}/.match(line.downcase) }
```

Comment:
My browser truncates the line after `match(lin` (123rd character). Please limit line length to about 100 for easier readability.